linked_list/linked_list.py

```diff
@@ -13,63 +13,140 @@ def __init__(self):
-        pass
+        self.head = Node(value, self.head)
 
     # method to find if the linked list contains a node with specified value
     # returns true if found, false otherwise
     # Time Complexity: ?
     # Space Complexity: ?
     def search(self, value):
-        pass
+        current = self.head 
+
+        while current != None:
+            if current.value == value:
+                return True 
+            current = current.next
+        return False 
 
     # method that returns the length of the singly linked list
     # Time Complexity: ?
     # Space Complexity: ?
     def length(self):
-        pass
+        current = self.head
+
+        count = 0 
+        while current != None:
+            current = current.next
+            count += 1
+        
+        return count
 
     # method that returns the value at a given index in the linked list
     # index count starts at 0
     # returns None if there are fewer nodes in the linked list than the index value
     # Time Complexity: ?
     # Space Complexity: ?
     def get_at_index(self, index):
-        pass
+        current = self.head
+
+        count = 0 
+        while current:
+            if count == index:
+                return current.value  
+            count += 1
+            current = current.next
+        
+        return None
+        
+    # method that returns the index of the first node with the given value
+    # returns -1 if the value is not found  in the linked list
+    # Time Complexity: O(n)?
+    # Space Complexity: O(1)?
+    
 
     # method that returns the value of the last node in the linked list
     # returns None if the linked list is empty
     # Time Complexity: ?
     # Space Complexity: ?
     def get_last(self):
-        pass
+
+        if self.head == None:
+            return None
+
+        current = self.head 
+
+        while current.next != None:
+            current = current.next
+
+        return current.value
 
     # method that inserts a given value as a new last node in the linked list
     # Time Complexity: ?
     # Space Complexity: ?
+    #method to add node to the end of the linked list
     def add_last(self, value):
-        pass
+        current = self.head
+        if current is None:
+            self.head = Node(value)
+            return
+        while current.next != None:
+            current = current.next
+        current.next = Node(value)
+        
 
     # method to return the max value in the linked list
     # returns the data value and not the node
     def find_max(self):
-        pass
+        if self.head == None:
+            return None
+
+        current = self.head
+
+        max = 0 
+        while current:
+            if current.value > max:
+                max = current.value 
+            current = current.next
+        
+        return max
 
     # method to delete the first node found with specified value
     # Time Complexity: ?
     # Space Complexity: ?

```

Comment:
What do you think the space and time complexity of this is? Is the time complexity dependent on the length of the linked list? Is there any new variables being created that grow larger depending upon the input?